mlir/lib/Conversion/SPIRVToLLVM/SPIRVToLLVM.cpp

```diff
@@ -260,12 +266,49 @@ static std::optional<Type> convertArrayType(spirv::ArrayType type,
   return LLVM::LLVMArrayType::get(llvmElementType, numElements);
 }
 
+#define STORAGE_SPACE_MAP(storage, space)                                      \
+  case spirv::StorageClass::storage:                                           \
+    return space;
+
+#define CLIENT_MAP(client, storage)                                            \
+  case spirv::ClientAPI::client:                                               \
+    return mapTo##client##AddressSpace(storage);
+
+static unsigned mapToOpenCLAddressSpace(spirv::StorageClass storageClass) {
+  // Based on clang/lib/Basic/Targets/SPIR.h.
+  switch (storageClass) {
+    STORAGE_SPACE_MAP(Function, 0)
+    STORAGE_SPACE_MAP(CrossWorkgroup, 1)
+    STORAGE_SPACE_MAP(Input, 1)
+    STORAGE_SPACE_MAP(UniformConstant, 2)
+    STORAGE_SPACE_MAP(Workgroup, 3)
+    STORAGE_SPACE_MAP(Generic, 4)
+    STORAGE_SPACE_MAP(DeviceOnlyINTEL, 5)
+    STORAGE_SPACE_MAP(HostOnlyINTEL, 6)
+  default:
+    return defaultAddressSpace;
+  }
+}
+
+static unsigned
+mapToAddressSpace(spirv::ClientAPI clientAPIForAddressSpaceMapping,
+                  spirv::StorageClass storageClass) {
+  switch (clientAPIForAddressSpaceMapping) {
+    CLIENT_MAP(OpenCL, storageClass)
```

Comment:
We should have a `Vulkan`, `WebGPU` and `Metal` case here. I'm not adding those because they're not relevant for us and I'd need to do some research to improve a feature we don't care about.

Reply:
OK then is fine to have the macro.